feat: Support 128-bit trace ids

fuzzybinary opened the pull request:
### What and why?

Adding support for 128-bit trace ids in DatadogTrackedWebRequest

Adding tests to ensure that the values generated for headers are the same as values generated for Datadog attributes.

refs: RUM-2757

### Review checklist

- [x] This pull request has appropriate unit and / or integration tests

fuzzybinary committed:
feat: Support 128-bit trace ids

Adding support for 128-bit trace ids in DatadogTrackedWebRequest

Adding tests to ensure that the values generated for headers are the same as values generated for Datadog attributes.

refs: RUM-2757

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/DatadogTrackedWebRequestTest.cs:
```diff
@@ -122,6 +123,91 @@ public void GeneratesCorrectDatadogAttributesWhenSampled()
+        [TestCase(TracingHeaderType.B3)]
+        [TestCase(TracingHeaderType.B3Multi)]
+        [TestCase(TracingHeaderType.TraceContext)]
+        public void DatadogAttributesAndTracingHeadersHaveSameValue(TracingHeaderType headerType)
```
/improvement

we should assert the content along with both attributes and tracing headers are same.

fuzzybinary replied:
Since this test generates a random `TraceContext` I don't know what the values should be.  I'll probably make two more tests that assert values of the Datadog Attributes are specific values and that Tracing Headers have specific values, which should test all three code paths.

fuzzybinary replied:
nm, I see what you're saying here. I've added lines that tests also test that the context value equals the attribute values.

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/DatadogTrackedWebRequestTest.cs:
```diff
@@ -122,6 +123,91 @@ public void GeneratesCorrectDatadogAttributesWhenSampled()
+            {
+                case TracingHeaderType.Datadog:
+                {
+                    BigInteger.TryParse(headers["x-datadog-trace-id"], out traceId);
+                    BigInteger.TryParse(headers["x-datadog-parent-id"], out spanId);
```
/improvement

you don't need big int for parsing here, datadog headers are still in the integer limits.

fuzzybinary replied:
Actually, since `GetIdsFromHeaders` is using `out BigInteger` it's easier to parse them here as BigInteger than to `ulong`. The performance is probably slightly worse but I'm not really concerned about it for the test.

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/TracingUUIDTest.cs:
```diff
@@ -38,6 +38,21 @@ public void TracingUuidDecRepresentationsAreCorrect(ulong high, ulong low, strin
+        public void TracingUuidLowDecRepresentationsAreCorrect(ulong high, ulong low, string expected)
+        {
+            // Given
+            var uuid = new TracingUuid(high, low);
```
/comment

let's keep the `TracingUuid` internal because this is not a UUID and we shouldn't communicate with type name its underlying representation.

fuzzybinary replied:
`TracungUuid` is already marked as `internal`.

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/TracingUUIDTest.cs:
```diff
@@ -126,11 +156,11 @@ public void CreateTracingUuid128BitSucceeds()
 
             // Then
             var str = uuid.ToString(TraceIdRepresentation.Hex);
-            Assert.Less(str.Length, 63);
+            Assert.LessOrEqual(str.Length, 32);
```
/improvement

should add Hex32Chars and Hex16Chars test case as well.

fuzzybinary replied:
Double checked and they already exist (line 89 and 74 respectively)

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/DatadogTrackedWebRequestTest.cs:
```diff
@@ -135,8 +221,19 @@ private static void VerifyHeaders(Dictionary<string, string> headers, TracingHea
+                        if (tagParts != null)
+                        {
+                            Assert.AreEqual(tagParts[0], "_dd.p.tid");
+                            BigInteger.TryParse(tagParts[1], NumberStyles.HexNumber, null, out var highTraceInt);
```
/improvement

This should be parsable via int.

fuzzybinary replied:
Yeah same as above, but parsable to `ulong` not `int`

fuzzybinary replied:
Actually, because this test is checking that the result fits in 63 bits, the use of `BigInteger.GetByteCount` simplifies it, so I'll probably leave it as is.

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/DatadogTrackedWebRequestTest.cs:
```diff
@@ -135,8 +221,19 @@ private static void VerifyHeaders(Dictionary<string, string> headers, TracingHea
+                        if (tagParts != null)
+                        {
+                            Assert.AreEqual(tagParts[0], "_dd.p.tid");
+                            BigInteger.TryParse(tagParts[1], NumberStyles.HexNumber, null, out var highTraceInt);
```
```suggestion
                            int.TryParse(tagParts[1], NumberStyles.HexNumber, null, out var highTraceInt);
```

ganeshnj commented on packages/Datadog.Unity/Tests/Rum/DatadogTrackedWebRequestTest.cs:
```diff
@@ -122,6 +123,91 @@ public void GeneratesCorrectDatadogAttributesWhenSampled()
+            {
+                case TracingHeaderType.Datadog:
+                {
+                    BigInteger.TryParse(headers["x-datadog-trace-id"], out traceId);
+                    BigInteger.TryParse(headers["x-datadog-parent-id"], out spanId);
```
```suggestion
                    int.TryParse(headers["x-datadog-trace-id"], out traceId);
                    int.TryParse(headers["x-datadog-parent-id"], out spanId);
```

fuzzybinary replied:
This would need to be `ulong` though, not int.

ganeshnj reviewed (commented):
Nothing major, small general improvement and better test coverage related.

Looks great.

fuzzybinary committed:
Add additional test assert

Add an assert that the values in the Datadog attributes are equal to what's in the tracing context.
